src/main/java/com/google/devtools/build/lib/remote/AbstractActionInputPrefetcher.java

```diff
@@ -548,11 +548,19 @@ public void finalizeAction(Action action, MetadataHandler metadataHandler) {
         inputsToDownload.add(output);
       }
 
-      for (Pattern pattern : patternsToDownload) {
-        if (pattern.matcher(output.getExecPathString()).matches()) {
-          outputsToDownload.add(output);
-          break;
+      if (output.isTreeArtifact()) {
+        var children = metadataHandler.getTreeArtifactChildren((SpecialArtifact) output);
+        if (outputMatchesPattern(output)) {
```

Comment:
If the pattern has a `$`, would this incorrectly match tree artifact paths?

Reply:
Do you have a concrete example?

If the pattern matches tree root, all files inside the tree are downloaded. (See newly added test cases)

Reply:
So that seems like a bug to me, since if we weren't using a tree artifact we would match download everything under `a/path` with the pattern `a/path$` (for example, if we had outputs `a/path/and/more/lib.a`.

Reply:
The check is guarded by `output.isTreeArtifact()`. File output `a/path/and/more/lib.a` won't match pattern `a/path$` so it won't be downloaded. Added another test case to illustrate that.

Reply:
Sure, but I'm saying that the behavior should differ between if it's a free artifact or not. We should match on the tree artifact itself if it has children, if using `$` in the pattern.

Reply:
(If I change the `lib.a` example from a normal output to a tree artifact output, the download behavior of _any_ patterns shouldn't change.)

Reply:
I see. Good catch! Updated the code.